src/components/lib/vmrt/vmrt.c

```diff
@@ -0,0 +1,427 @@
+	assert(!vcpu->shared_region);
+
+	vcpu->vm = vm;
+
+	handler_tid = sched_thd_create(vmrt_vm_exception_handler, vcpu);
+	vcpu->handler_tid = handler_tid;
+
+	vcpu->vpid = vcpu_nr + 1;
+	vmrt_vmcs_page_create(vcpu);
+	vmrt_msr_bitmap_page_create(vcpu);
+	vmrt_lapic_page_create(vcpu);
+	vmrt_shared_page_create(vcpu);
+	vmrt_lapic_access_page_create(vcpu);
+	vmrt_vmcb_create(vcpu);
+
+	cap = capmgr_vm_vcpu_create(vcpu->vm->comp_id, vcpu->vmcb_cap, &tid);
+	assert(tid != 0);
+	vcpu->tid = tid;
+	vcpu->cap = cap;
+	vcpu->cpuid = vcpu_nr;
+	vcpu->coreid = cos_coreid();
+
+	vcpu->next_timer = ~0ULL;
+
+}
+
+void
+vmrt_vm_mem_init(struct vmrt_vm_comp *vm, void *mem)
+{
+	vm->guest_addr = mem;
+
+	/* Clean the memory for the VM */
+	memset(mem, 0, vm->guest_mem_sz);
+}
+
+void
+vmrt_vm_create(struct vmrt_vm_comp *vm, char *name, u8_t num_vcpu, u64_t vm_mem_sz)
+{
+	compid_t vm_id = 0;
+	
+	vaddr_t lapic_access_page;
+
+	assert((vm_mem_sz % PAGE_SIZE_4K) == 0);
+
+	vm_id = capmgr_vm_comp_create(vm_mem_sz);
+	vm->comp_id = vm_id;
+
+	/*
+	 * Allocate lapic access page for the VM.
+	 *
+	 * This is to allocate the "lapic access page" for the VM.
+	 * Simply, this is just to allocate a kernel page, mapped it
+	 * into VM comp's EPT at a specified address, by default, it
+	 * is LAPIC_BASE_ADDR. Then, this page needs to be also referenced
+	 * by the VMCS in the kernel.
+	 * 
+	 * This page is only used by a VMCS to let the vcpu know
+	 * it is accessing a lapic page. The real lapic page used by each vcpu
+	 * to read/write data will be allocated later when initializing vcpu.
+	 */
+	lapic_access_page = capmgr_vm_shared_kernel_page_create_at(vm_id, LAPIC_BASE_ADDR);
+	assert(lapic_access_page);
+	vm->lapic_access_page = lapic_access_page;
+
+	assert(num_vcpu < VMRT_VM_MAX_VCPU);
+	vm->num_vpu = num_vcpu;
+	vm->guest_mem_sz = vm_mem_sz;  
+
+	strcpy((char *)&vm->name, name);
+
+	return;
+}
+
+void
+vmrt_vm_data_copy_to(struct vmrt_vm_comp *vm, char *data, u64_t size, paddr_t gpa)
+{
+	char *host_va = GPA2HVA(gpa, vm);
+	memcpy(host_va, data, size);
+}
+
+void
+vmrt_vm_vcpu_resume(struct vmrt_vm_vcpu *vcpu)
+{
+	cos_switch(vcpu->cap, 0, 0, 0, 0, 0);
+}
+
+
+CWEAKSYMB void 
+rdmsr_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	VM_PANIC(vcpu);
+}
+
+CWEAKSYMB void 
+wrmsr_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	VM_PANIC(vcpu);;
+}
+
+CWEAKSYMB void 
+ept_misconfig_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	VM_PANIC(vcpu);
+}
+
+CWEAKSYMB void 
+vmcall_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	VM_PANIC(vcpu);
+}
+
+CWEAKSYMB void 
+xsetbv_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	/* Note: the host uses the same xcr0 as guest, thus this can be bypassed */
+	GOTO_NEXT_INST(vcpu->shared_region);
+}
+
+CWEAKSYMB void
+lapic_write_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	VM_PANIC(vcpu);
+}
+
+CWEAKSYMB void 
+io_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	VM_PANIC(vcpu);
+}
+
+CWEAKSYMB void 
+cpuid_handler(struct vmrt_vm_vcpu *vcpu)
+{
+	VM_PANIC(vcpu);
+}
+
+
+CWEAKSYMB void 
```

Comment:
If you took any code from external sources, we need to attribute it, and make sure that the licenses match up.